OrbitLinuxTracing/ThreadStateVisitor.cpp

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ThreadStateVisitor.h"
+
+namespace LinuxTracing {
+
+using orbit_grpc_protos::ThreadStateSlice;
+
+// Note: Since we use PerfEventProcessor to process perf_event_open events in order, OnNewTask,
+// OnSchedWakeup, OnContextSwitchIn, OnContextSwitchOut are expected to be called in order of
+// timestamp. But the initial thread states are retrieved (and OnInitialState is called) after the
+// perf_event_open file descriptors have been enabled, so that we don't lose thread states between
+// retrieving the initial states and enabling the file descriptors. It is then common for some of
+// the first tracepoint events to have a timestamp lower than the timestamp of initial retrieval. In
+// all these cases, we discard the previous known state (the one retrieved at the beginning, with a
+// larger timestamp) and replace it with the thread state carried by the tracepoint.
+
+void ThreadStateManager::OnInitialState(uint64_t timestamp_ns, pid_t tid,
```

Comment:
Feels a bit strange that the ThreadStateManager and the ThreadStateVisitor are in the same file? Could you separate them?

Reply:
I see your point. But couldn't we then make it to a private class and let the test be a friend class?
Anyways. I don't have a strong opinion on that. I personally would separate this (mostly for readability). Probably going the friend class path, would be a bit of an overkill. So feel free to ignore.